lib/core/networking/websockets/middleware.dart

```diff
@@ -0,0 +1,83 @@
+import '../protocol/authentication.dart';
+import '../protocol/protocol.dart';
+
+import 'session.dart';
+
+//* Type declarations
+/// Different actions middleware can handle
+enum MiddlewareAction {
+  recieve,
+  send,
+}
+
+/// A Middleware function for use in MiddlewareSession
+///
+/// Arguments: Calling [Session] instance, message, [MiddlewareAction] action and a proxy to the next [MiddlewareFunc].
+typedef MiddlewareFunc = Future<dynamic> Function(
+  dynamic,
```

Comment:
Because we dont want to cast `ServerSession`/`ClientSession` to `Session`.

Reply:
You don't have to. You can pass a `ServerSession` as a `Session` object because `ServerSession` extends `Session`. If you type check the object will still be `ServerSession`.

```dart
void doOnSession(Session obj) {
  if (obj is ServerSession) {
    print('This is a server session and we can use all it's custom properties and methods. Also overrides will work as expected')
  }
```

Reply:
Amongst other issues, this would also break our tests, as `MockSession` is not derived from `Session<T>`. I would like to stay with `dynamic`.

Reply:
Wouldn't it be the much cleaner solution to derive `MockSession` from `Session` and expect a `Session` as written in the original post? When you are expecting in production to receive always a `Session` why not test it with `Session`?

Reply:
Because we want to test middleware without the whole networking stack (running server and client)